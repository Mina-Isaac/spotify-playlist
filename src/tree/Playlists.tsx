import { types, flow, Instance } from "mobx-state-tree";
import appService from "./service";

interface IPlaylist extends Instance<typeof playlist> {}
export interface IPlaylists extends Instance<typeof Playlists> {}

const image = types.model({
  height: types.union(types.number, types.null),
  width: types.union(types.number, types.null),
  url: types.string,
});

const owner = types.model({
  display_name: types.string,
});

const playlist = types.model({
  id: types.identifier,
  name: types.string,
  description: types.string,
  images: types.array(image),
  owner,
});

const Playlists = types
  .model({
    items: types.array(playlist),
    next: types.union(types.string, types.null),
    total: types.number,
    loading: types.boolean,
  })
  .actions((self) => {
    const fetchNext = flow(function* () {
      if (self.next === null) return;
      self.loading = true;
      try {
        const newlyFetched = yield appService.fetchPlaylists(self.next);
        self.next = newlyFetched.next;
        self.loading = false;
        (newlyFetched.items as Array<IPlaylist>).forEach((item) =>
          self.items.push(item)
        );
      } catch (error) {
        console.error("Failed to fetch projects", error);
        self.loading = false;
      }
      return;
    });
    return { fetchNext };
  });

export default Playlists;
