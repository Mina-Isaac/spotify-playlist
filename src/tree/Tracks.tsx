import { types, flow, Instance } from "mobx-state-tree";
import appService from "./service";

interface ITrack extends Instance<typeof track> {}

const artist = types.model({
  id: types.identifier,
  name: types.string,
  type: types.string,
});

const track = types.model({
  track: types.model({
    artists: types.array(artist),
  }),
});

const Tracks = types
  .model({
    items: types.array(track),
    next: types.union(types.string, types.null),
    total: types.number,
    loading: types.boolean,
  })
  .actions((self) => {
    const fetchNext = flow(function* (id?) {
      if (self.next === null) return;
      self.loading = true;
      try {
        const newlyFetched = yield appService.fetchTracks(id, self.next);
        self.next = newlyFetched.next;
        (newlyFetched.items as Array<ITrack>).forEach((item) =>
          self.items.push(item)
        );
        self.loading = false;
      } catch (error) {
        console.error("Failed to fetch projects", error);
        self.loading = false;
      }
      return;
    });
    const clearTracks = () => {
      self.items.clear();
      self.loading = true;
      self.next = "";
      self.total = 0;
    };
    return { fetchNext, clearTracks };
  });

export default Tracks;
