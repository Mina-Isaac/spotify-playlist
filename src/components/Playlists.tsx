import React, { useEffect } from "react";
import useInfiniteScroll from "../utilities/useInfiniteScroll";
import { observer } from "mobx-react";
import { useMst } from "../tree/Root";
import { Link } from "react-router-dom";

const App = () => {
  const root = useMst();
  const list = root.Playlists.items;
  const setIsFetching = useInfiniteScroll(moreData)[1]
  const loading = root.Playlists.loading;

  function moreData() {
    root.Playlists.fetchNext();
    setIsFetching(false);
  }

  useEffect(() => {
    root.Playlists.fetchNext();
  }, []);

  if (list.length === 0) {
    return <h1>Loading...</h1>;
  }

  return (
    <>
      <h1>List of featured playlists</h1>
      <ul className="list">
        {list.map((item, i) => (
          <li className="card" key={i}>
            <Link
              className="link"
              to={{
                pathname: `/spotify-playlist/artists/${item.id}`,
                state: item.name,
              }}
            >
              <img
                width={"120"}
                src={item.images[0].url}
                alt={item.description}
              />
              <div className="txt">
                <span className="primary-txt">{item.name}</span>
                <span className="secondery-txt">{`by ${item.owner.display_name}`}</span>
              </div>
            </Link>
          </li>
        ))}
      </ul>
      {loading && <span className='primary-text'>Loading...</span> }
    </>
  );
};

export default observer(App);
