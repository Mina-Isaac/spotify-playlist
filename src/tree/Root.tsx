import { types, Instance } from "mobx-state-tree";
import Playlists from "./Playlists";
import Tracks from "./Tracks";
import { createContext, useContext } from "react";

export interface IRoot extends Instance<typeof Root> {}
//export type RootInstance = Instance<typeof Root>;

const Root = types.model({
  Playlists,
  Tracks,
});

export const store = Root.create({
  Playlists: {
    items: [],
    next: "",
    total: 0,
    loading: false,
  },
  Tracks: {
    items: [],
    next: "",
    total: 0,
    loading: false,
  },
});

const RootStoreContext = createContext<null | IRoot>(null);
export const { Provider } = RootStoreContext;
export function useMst() {
  const store = useContext(RootStoreContext);
  if (store === null) {
    throw new Error("Store cannot be null, please add a context provider");
  }
  return store;
}
