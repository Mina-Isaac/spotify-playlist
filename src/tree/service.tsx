import { PLAYLIST_ENDPOINT, TRACKS_ENDPOINT, ACCESS_TOKEN } from "../Constants";
import { IPlaylists } from './Playlists'

const appService = {
    
  fetchPlaylists(next: string): Promise<any> {
    const url = next || `${PLAYLIST_ENDPOINT}fields=playlists.items(id,images,name,owner),next,total&&limit=30&&offset=0`
    return fetch(url, {
      headers: {
        "Content-Type": "application/json",
        Authorization:
          `Bearer ${ACCESS_TOKEN}`,
      },
    })
      .then((res) => res.json())
      .then((res) => res.playlists as IPlaylists);
  },

  fetchTracks(id: string, next: string): Promise<any> {
    const url =next || `${TRACKS_ENDPOINT + id}/tracks?fields=items(track.artists),next,total&&limit=30&&offset=0`
    return fetch(url, {
      headers: {
        "Content-Type": "application/json",
        Authorization:
          `Bearer ${ACCESS_TOKEN}`,
      },
    })
      .then((res) => res.json())
  },
};

export default appService;
