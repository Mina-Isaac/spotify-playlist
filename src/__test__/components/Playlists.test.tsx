import React from "react";
import ReactDOM from "react-dom";
import Playlists from "../../components/Playlists";
import renderer from "react-test-renderer";
import { store, Provider } from "../../tree/Root";

describe("Playlists component should render without issues", () => {
  beforeEach(() => {});
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider value={store}>
        <Playlists />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it("component renders correctly", () => {
    const tree = renderer
      .create(
        <Provider value={store}>
          <Playlists />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
