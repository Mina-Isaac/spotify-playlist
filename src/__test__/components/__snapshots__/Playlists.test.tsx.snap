// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlists component should render without issues component renders correctly 1`] = `
<h1>
  Loading...
</h1>
`;
